lib/interface/cli/commands/offline-logs/offload-to-collection.cmd.js

```diff
@@ -1,96 +1,278 @@
-  
-  const cutoffDateObj = moment(cutoffDate)
-                        .add(1, 'days')
-                        .startOf("day")
-
-  if(!cutoffDateObj.isValid()){
-    throw new Error('please enter a valid date in ISO 8601 format')
+const moment = require("moment");
+const semver = require("semver");
+const Command = require("../../Command");
+const cmd = require("./base.cmd");
+const {
+  objectIdFromDate,
+  defaultCollections,
+  collectionExists,
+  findMinLog,
+} = require("./utils");
+
+const offloadToCollection = async function (
+  sourceDBObj,
+  collection,
+  targetDB,
+  cutoffDate
+) {
+  const sourceCollectionObj = sourceDBObj.collection(collection);
+  const targetCollection = `archive-${collection}`;
+
+  const cutoffDateObj = moment(cutoffDate).add(1, "days").startOf("day");
+
+  if (!cutoffDateObj.isValid()) {
+    throw new Error("please enter a valid date in ISO 8601 format");
+  }
+
+  const cutoffDateId = objectIdFromDate(cutoffDateObj.toDate());
+
+  const mongoInfo = await sourceDBObj.admin().serverInfo();
+  const mongoServerVersion = await mongoInfo.version;
+  const mergeReleaseVersion = "4.2.0";
+
+  if (semver.lte(mergeReleaseVersion, mongoServerVersion)) {
+    await archiveWithMerge(
+      sourceCollectionObj,
+      cutoffDateId,
+      targetDB,
+      targetCollection,
+      collection
+    );
+  } else {
+    if (await collectionExists(sourceDBObj, targetCollection)) {
+      await archiveInChunks(
+        sourceDBObj,
+        sourceCollectionObj,
+        targetCollection,
+        cutoffDateId,
+        collection,
+        cutoffDateObj
+      );
+    } else {
+      await archiveWithOut(
+        sourceCollectionObj,
+        cutoffDateId,
+        targetCollection,
+        collection
+      );
+    }
+  }
+};
+///////////////////////////////////////////////////////////////////////////////////
+const archiveWithMerge = async function (
+  sourceCollectionObj,
+  cutoffDateId,
+  targetDB,
+  targetCollection,
+  collection
+) {
+  const minLog = await findMinLog(
+    sourceCollectionObj,
+    cutoffDateId,
+    collection
+  );
+
+  if (!minLog) {
+    console.info(
+      `No logs to archive in Collection ${collection} from the given date.`
+    );
+    return;
+  }
+
+  var result = sourceCollectionObj.aggregate([
+    { $match: { _id: { $lte: ObjectId(cutoffDateId) } } },
+    {
+      $merge: {
+        into: { db: targetDB, coll: targetCollection },
+        on: "_id",
+        whenMatched: "keepExisting",
+        whenNotMatched: "insert",
+      },
+    },
+  ]);
+
+  await result.toArray();
+
+  if (!result.cursorState.killed) {
+    console.info(
+      `logs from '${collection}' were archived to '${targetCollection}'`
+    );
+    await sourceCollectionObj.deleteMany({
+      _id: { $lte: ObjectId(cutoffDateId) },
+    });
+  } else {
+    throw new Error(
+      "Cursor error. Archiving operation may not be completed.\
+    The old logs were not deleted from the source collection."
+    );
```

Comment:
also - this bit can be extracted into some "checkCursorState" func

Reply:
done